server/service/src/programs/contact_trace/contact_trace_updated.rs

```diff
@@ -0,0 +1,90 @@
+use super::{
+    contact_trace_schema::{SchemaContactTrace, SchemaGender},
+    upsert::UpsertContactTraceError,
+};
+use chrono::{DateTime, NaiveDate};
+use repository::{
+    contact_trace::{ContactTraceFilter, ContactTraceRepository},
+    contact_trace_row::{ContactTraceRow, ContactTraceRowRepository},
+    Document, Gender, ProgramRow, StorageConnection, StringFilter,
+};
+use std::str::FromStr;
+use util::hash::sha256;
+
+/// Callback called when a program enrolment document has been updated
+pub(crate) fn update_contact_trace_row(
+    con: &StorageConnection,
+    patient_id: &str,
+    document: &Document,
+    contact_trace: SchemaContactTrace,
+    program_row: ProgramRow,
+) -> Result<ContactTraceRow, UpsertContactTraceError> {
+    let datetime = DateTime::parse_from_rfc3339(&contact_trace.datetime)
+        .map_err(|err| {
+            UpsertContactTraceError::InternalError(format!(
+                "Invalid enrolment datetime format: {}",
+                err
+            ))
+        })?
+        .naive_utc();
+
+    // Retrieve existing or create now contact trace id
+    let repo = ContactTraceRepository::new(con);
+    let contact_trace_row = repo
+        .query_by_filter(ContactTraceFilter {
+            document_name: Some(StringFilter::equal_to(&document.name)),
+            ..ContactTraceFilter::default()
+        })?
+        .pop();
+    let contact_trace_id = match contact_trace_row {
+        Some(contact_trace_row) => contact_trace_row.0.id,
+        None => sha256(&document.name),
+    };
+
+    let store_id = contact_trace
+        .location
+        .as_ref()
+        .and_then(|l| l.store_id.clone());
+    let contact = contact_trace.contact.as_ref();
+
+    let row = ContactTraceRow {
+        id: contact_trace_id,
+        program_id: program_row.id,
+        document_id: document.id.clone(),
+        patient_id: patient_id.to_string(),
+        datetime,
+        contact_trace_id: contact_trace.contact_trace_id,
+        contact_patient_id: contact.and_then(|c| c.id.clone()),
+        first_name: contact.and_then(|c| c.first_name.clone()),
```

Comment:
They might all not be known or not be recorded... (These are the details of the contact given by the patient, e.g. patient says something like "There was this other person I don't really know, but when I run into him, I will send him over.")

Reply:
Would the patient be given a contact trace id for them to write down and give to that person ?

Reply:
I don't believe this is currently the case. How a contact is matched to a record wasn't clear to us in the beginning. But it seems that the limited information that is recorded is enough to match the contact with an existing entry (I guess because there is only a limited/small number of contacts for each patient)... We added fields to record the contact's name which should make it easier to match a contact.